grpc/src/main/java/com/linecorp/armeria/server/grpc/HttpJsonTranscodingService.java

```diff
@@ -416,6 +426,61 @@ private static Descriptor findTypeDescriptor(FileDescriptor file, FieldDescripto
         return file.findMessageTypeByName(messageType.getName());
     }
 
+    // to make it more efficient, we calculate whether extract response body one time
+    // if there is no matching toplevel field, we set it to null
+    @Nullable
+    private static String validateResponseBody(List<FieldDescriptor> topLevelFields,
+                                               @Nullable String responseBody) {
+        if (StringUtil.isNullOrEmpty(responseBody)) {
+            return null;
+        }
+        for (FieldDescriptor fieldDescriptor: topLevelFields) {
+            if (fieldDescriptor.getName().equals(responseBody)) {
+                return responseBody;
+            }
+        }
+        return null;
+    }
+
+    @Nullable
+    private static Function<HttpData, HttpData> generateResponseBodyConverter(TranscodingSpec spec) {
+        @Nullable final String responseBody = spec.responseBody;
+        if (responseBody == null) {
+            return null;
+        } else {
+            return httpData -> {
+                try (HttpData data = httpData) {
+                    final byte[] array = data.array();
+                    try {
+                        final JsonNode jsonNode = mapper.readValue(array, JsonNode.class);
+                        // we try to convert lower snake case response body to camel case
+                        final String lowerCamelCaseResponseBody =
+                                CaseFormat.LOWER_UNDERSCORE.to(CaseFormat.LOWER_CAMEL, responseBody);
+                        final Iterator<Entry<String, JsonNode>> fields = jsonNode.fields();
+                        while (fields.hasNext()) {
+                            final Entry<String, JsonNode> entry = fields.next();
+                            final String fieldName = entry.getKey();
+                            final JsonNode responseBodyJsonNode = entry.getValue();
+                            // try to match field name and response body
+                            // 1. by default the marshaller would use lowerCamelCase in json field
+                            // 2. when the marshaller use original name in .proto file when serializing messages
+                            if (fieldName.equals(lowerCamelCaseResponseBody) ||
+                                fieldName.equals(responseBody)) {
+                                final byte[] bytes = mapper.writeValueAsBytes(responseBodyJsonNode);
+                                return HttpData.wrap(bytes);
+                            }
+                        }
+                        return HttpData.ofUtf8("null");
```

Comment:
Question, what do you think of returning an empty json body instead of the string `null`? I guess most clients will try to parse a json response body

```suggestion
                        return HttpData.ofUtf8("{}");
```

Reply:
Isn’t `null` valid JSON? I thought it meant nothing was filtered by responseBody.

Reply:
Ah it seems so https://www.rfc-editor.org/rfc/rfc7159#section-3.  TIL 👍